model/src/component/button.rs

```diff
@@ -0,0 +1,144 @@
+use crate::component::Component;
+use crate::id::EmojiId;
+use serde::{Deserialize, Serialize};
+use serde_repr::{Deserialize_repr, Serialize_repr};
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, Hash)]
+pub struct Button {
+    pub style: ButtonStyle,
+    pub emoji: Option<PartialEmoji>,
+    pub label: Option<String>,
+    pub custom_id: Option<String>,
+    pub url: Option<String>,
+    pub disabled: bool,
+}
+
+#[derive(
+    Clone, Copy, Debug, Deserialize_repr, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize_repr,
+)]
+#[repr(u8)]
+pub enum ButtonStyle {
+    Primary = 1,
+    Secondary = 2,
+    Success = 3,
+    Danger = 4,
+    Link = 5,
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, Hash)]
+pub struct PartialEmoji {
+    pub id: Option<EmojiId>,
+    pub name: String,
+    pub animated: bool,
+}
+
+/// Helper struct for assembling  the Button struct
+#[derive(Debug)]
+pub struct ButtonBuilder {
+    pub(crate) style: ButtonStyle,
+    pub(crate) emoji: Option<PartialEmoji>,
+    pub(crate) label: Option<String>,
+    pub(crate) custom_id: Option<String>,
+    pub(crate) url: Option<String>,
+    pub(crate) disabled: bool,
+}
+
+#[derive(Debug)]
+pub enum ButtonBuilderError {
+    MissingLabelOrEmoji,
+    LabelTooLong,
+    CustomIdTooLong,
+    LinkTooLong,
+}
+
+impl ButtonBuilder {
+    /// Makes a new button that links to an external site
+    /// You will also need to set the emoji or label as well (or both)
+    ///
+    /// # Errors
+    /// Emits `ButtonBuilderError::LinkTooLong` if the link is over 512 chars
```

Comment:
Please format these docs similarly to other docs: One line for a short description, followed by lines about other things. There needs to be a newline after the error section. The Error type needs to link to the error itself.